### ISSUE TITLE
On-device sleep stage classifier for real-time temperature autopilot

### ISSUE THREAD

**Issue by ng** [sleepypod/core#198]
## Summary

Add a 5th Python module (`sleep-classifier`) that runs on the pod and classifies sleep stage in real-time (Wake/Light/Deep/REM) using BCG-derived signals. This enables autopilot temperature control — adjusting the SleepCurve based on actual sleep state rather than fixed time-based setpoints.

## Motivation

The current temperature curve (`SleepCurve.swift`) is open-loop and time-based. The pod's thermal system has ~15-20 min lag per 1°F change. A closed-loop system that knows "the user just entered deep sleep" can:
- Hold cooling longer if deep sleep starts early
- Warm slightly during restless periods to stabilize light sleep
- Pause thermal control during bathroom trips
- Extend the cold hold if N3 is still detected past the scheduled ramp-up

## Available signals (all on-pod today)

| Signal | Source | Table | Rate |
|---|---|---|---|
| Heart rate | piezo-processor | `vitals` | ~60s |
| HRV (RMSSD) | piezo-processor | `vitals` | ~60s |
| Breathing rate | piezo-processor | `vitals` | ~60s |
| Presence | sleep-detector (capSense/capSense2) | `sleep_records` | ~1s |
| Movement | sleep-detector | `movement` | ~60s |
| Bed temperature | environment-monitor | `bed_temp` | ~15s |

### Phase 1: Rule-based (ship immediately, ~60-65% accuracy)

4-class: Wake / Light (N1+N2) / Deep (N3) / REM. N1 and N2 are indistinguishable without EEG.

**Personal baseline**: compute mean HR and BR over minutes 30-60 after bed entry (first stable sleep period). All thresholds are relative to this baseline.

#### Threshold table (from literature synthesis)

| Signal | Wake | Light (N1/N2) | Deep (N3) | REM |
|---|---|---|---|---|
| HR vs nightly mean | Above mean | Near/slightly below | **8-15% below** (nightly min) | Returns toward mean |
| HR epoch variability | High | Moderate | **Low** (stable) | **High** (surges) |
| RMSSD (5-min) | Moderate | Increasing | **Highest** (peak vagal) | Drops below NREM |
| Breathing rate | 14-18 brpm | 12-16 brpm | 10-15 brpm | 12-16 brpm |
| Breathing regularity | Low | High | **Highest** | **Low** (irregular) |
| Movement density | High | Low | **Near-zero** | Near-zero |

**Decision logic (priority order):**
1. Movement above threshold → **Wake**
2. No presence detected → **Wake**
3. HR > nightly mean AND high movement density → **Wake**
4. HR < 85% of nightly mean AND RMSSD elevated AND breathing regular AND movement near-zero → **Deep**
5. HR near mean AND RMSSD low AND breathing irregular AND movement near-zero AND elapsed > 60 min → **REM**
6. Else → **Light**

#### Post-processing smoothing rules

From Ibrahim et al. 2019 and sleep architecture literature:

- **REM latency gate**: No REM classification before 60 min after sleep onset (reclassify as Light)
- **Forbidden transitions**: Wake→REM, Wake→Deep, N3→REM directly (must pass through Light)
- **Epoch smoothing**: For 3 consecutive epochs A,B,C — if A==C and B≠A and B≠Wake, reclassify B as A
- **Minimum stage duration**: REM episodes < 2 min → reclassify as Light
- **Architecture priors**: N3 predominates first 1/3 of night; REM predominates last 1/3

### Phase 2: Simple ML (~70-78% accuracy)

Random Forest or gradient-boosted trees with engineered features:

**Feature set per 30-second epoch:**
- HR: mean, z-score vs nightly mean, slope (5-min trailing), CV
- HRV: 5-min RMSSD, SDNN, pNN50, HF power, LF/HF ratio
- BR: mean, z-score vs nightly mean, breath-to-breath CV, spectral entropy
- Movement: density (fraction of 1s sub-epochs with motion), intensity (RMS), time since last movement
- Temporal: minutes since sleep onset, elapsed fraction of expected sleep, clock hour (circadian proxy)

**Training data**: pair pod vitals with Apple Watch sleep stages via HealthKit (per iOS ADR-004). Apple Watch provides PSG-validated 4-class labels.

### Phase 3: Deep learning (~77-80% accuracy)

LSTM or 1D-CNN on raw feature sequences. Reference: Radha/Fonseca 2019 achieved 77% accuracy, kappa 0.61 on HRV alone with LSTM. SleepNetZero 2024 achieved 80.3% kappa 0.718 on BCG specifically.

## Architecture

```
modules/
├── sleep-classifier/
│   ├── main.py                 # main loop: poll vitals, classify, write, optionally adjust temp
│   ├── classifier.py           # rule-based stage detection (Phase 1)
│   ├── features.py             # feature extraction from vitals/movement windows
│   ├── smoothing.py            # post-processing (architecture constraints, transition rules)
│   ├── autopilot.py            # temperature adjustment via DAC socket (optional, gated)
│   ├── manifest.json
│   ├── requirements.txt
│   └── sleepypod-sleep-classifier.service
```

**New DB table:**
```sql
CREATE TABLE sleep_stages (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    side TEXT NOT NULL,           -- 'left' / 'right'
    timestamp INTEGER NOT NULL,  -- unix epoch
    stage TEXT NOT NULL,          -- 'wake' / 'light' / 'deep' / 'rem'
    confidence REAL,             -- 0.0-1.0
    hr_baseline REAL,            -- personal nightly mean HR used
    features TEXT,               -- JSON of feature values (for debugging/training)
    UNIQUE(side, timestamp)
);
```

## Autopilot integration

When autopilot mode is enabled via the iOS app:

1. Classifier writes current stage to `sleep_stages`
2. Autopilot reads stage + current SleepCurve schedule
3. Adjusts temperature target:
   - Deep detected early → hold cold setpoint (don't ramp up yet)
   - Restless/Wake → warm 1-2°F to stabilize
   - REM detected → hold current temp (REM needs stable thermal environment)
   - Out of bed > 2 min → pause heating/cooling
4. Writes adjustment to DAC via hardware client
5. Logs decision to `autopilot_log` table for morning review

## Safety bounds

- All temperature adjustments clamped to 55-110°F (existing bounds)
- Autopilot adjustments limited to ±3°F from the scheduled setpoint (prevents runaway)
- Autopilot disengages if classifier confidence drops below 0.3 for > 5 min
- Kill switch via iOS app or tRPC `autopilot.disable()`

## Overnight logging (for validation before enabling autopilot)

Run classifier in **observe-only mode** first — classify and log but don't adjust temperature. Review in the morning via tRPC `sleepStages.getHistory()` or iOS visualization. Compare against subjective experience and (if available) Apple Watch data.

## Key references

- Fonseca et al. 2015 — ECG+respiratory sleep staging, 142 features, 69% accuracy
- Radha/Fonseca et al. 2019 — LSTM on HRV, 77% accuracy, kappa 0.61
- Walch et al. 2019 — consumer device staging, ~65% accuracy
- SleepNetZero (Li et al. 2024) — BCG-specific deep learning, 80.3% accuracy
- Heise et al. 2022 — BCG SVM, 85.5% accuracy (3-class)
- CReSS (Yetton et al. 2021) — cardiorespiratory staging, 78% accuracy, kappa 0.643
- Penzel et al. 2003 — HR dynamics across sleep stages
- Trinder et al. 2001 — autonomic activity during sleep (HRV 123.7% higher in SWS vs REM)
- Douglas et al. 1982 — respiratory patterns during sleep

## Depends on

- #195 capSense2 support (merged) — presence detection on Pod 5
- #196 piezo quality scoring fix — ensures reliable vitals data
- Calibrated sensors — classifier accuracy depends on clean baselines
